client/finality-grandpa/src/environment.rs

```diff
@@ -1174,72 +1174,80 @@ where
 	let limit = authority_set.current_limit(*base_header.number());
 	debug!(target: "afg", "Finding best chain containing block {:?} with number limit {:?}", block, limit);
 
-	let result = match select_chain.finality_target(block, None).await {
-		Ok(best_hash) => {
-			let best_header = client
-				.header(BlockId::Hash(best_hash))?
-				.expect("Header known to exist after `finality_target` call; qed");
+	let mut target_header = match select_chain.finality_target(block, None).await {
+		Ok(target_hash) => client
+			.header(BlockId::Hash(target_hash))?
+			.expect("Header known to exist after `finality_target` call; qed"),
+		Err(err) => {
+			warn!(target: "afg",
+				"Encountered error finding best chain containing {:?}: couldn't find target block: {}",
+				block,
+				err,
+			);
 
-			// check if our vote is currently being limited due to a pending change
-			let limit = limit.filter(|limit| limit < best_header.number());
+			return Ok(None)
+		},
+	};
 
-			let (base_header, best_header, target_header) = if let Some(target_number) = limit {
-				let mut target_header = best_header.clone();
+	// NOTE: this is purposefully done after `finality_target` to prevent a case
+	// where in-between these two requests there is a block import and
+	// `finality_target` returns something higher than `best_chain`.
+	let best_header = match select_chain.best_chain().await {
```

Comment:
dumb question: can `best_chain` legitimately change after `finality_target` is called for other reasons than block import? 
For example, if we enable stagnant checks and a stagnant branch become viable. 
Or a disputes resolves and we do a reversion or something along the lines.

Reply:
Yes, I guess it could change in-between for those two reasons you mentioned. Maybe the API of `SelectChain` should be changed so that we can return both blocks in a single call, hence ensuring that they are consistent with each other. I'll do that in a follow-up PR (it will require changes in Polkadot as well).